Use an internal buffer to increase _changes throughput

kocolosk opened the pull request:
This set of commits refactors the `changes_callback` function to keep an internal data buffer and flush it out the socket after a configurable number of bytes have been accumulated. It also adds support for a new callback message that will cause it to flush the buffer at the end of each traversal of the sequence tree for "live" feeds.

COUCHDB-2724

kocolosk committed:
Remove temporary upgrade clause

This clause was only needed for a very old hot code upgrade.

kocolosk committed:
Use a record for changes_callback accumulator

This change allows us to evolve the accumulator in a less-brittle way
and sets the stage for new data to be held in the accumulator to address
COUCHDB-2724.

In the course of this change I also switched the feed labels from strings
to atoms (they're only used for pattern matching in the accumulator, and
multiple matches are executed for every row in the feed, so it seemed
silly to be using Erlang lists for that comparison), and I explicitly
indicated when we start a chunked response instead of guessing it
heuristically based on other contents in the accumulator.

COUCHDB-2724

kocolosk committed:
Buffer rows for normal/longpoll feeds

This patch causes the coordinator to accumulate data in its own buffer
and reduce the number of calls to write data on the socket. The size of
the buffer is configurable:

[httpd]
chunked_response_buffer = 1490

The default is chosen to approximately fill a standard Ethernet frame.

COUCHDB-2724

kocolosk committed:
Add basic buffering support for other feed types

With this code it is possible that changes are buffered for a long
period of time and not sent out. Will work on addressing that next.

COUCHDB-2724

kocolosk committed:
Execute a callback for every complete DB traversal

This ensures that we don't enter a receive statement waiting for new DB
updates without first flushing the buffer.

COUCHDB-2724

kxepal commented:
Nice! How much does this improves throughput?

kxepal commented on src/chttpd_db.erl:
```diff
@@ -74,6 +86,8 @@ handle_changes_req1(#httpd{}=Req, Db) ->
         filter_fun = couch_changes:configure_filter(Raw, Style, Req, Db)
     },
+    % Default to ~filling the payload of a standard Ethernet frame
+    Max = config:get_integer("httpd", "chunked_response_buffer", 1490),
```
I believe, this option we should advice to tweak if user iface is configured for some other MTU, right? Also, standard Ethernet frame is 1500, why 1490 had chosen here?

kocolosk replied:
We should figure out what the optimal size really is; filling an Ethernet frame was almost an arbitrary choice on my part. We did see a larger increase in throughput with a 64KB buffer. If the frame size really does turn out to be a good target then yes, we should either programmatically discover the MTU or document the way to optimize this.

I chose 1490 instead of 1500 because mochiweb needs a couple of bytes to declare the size of the chunk and to add a newline or two. I didn't compute exactly how many bytes are required for that "envelope"; it's probably a bit less than 10 but I figured it was better to err on the conservative side than to create a new, nearly-empty packet.

kocolosk commented:
The speedup depends on how CPU-constrained you are; as I mentioned in the JIRA I constructed a test on my laptop where setting the buffer to 64KB resulted in about a 60% improvement in throughput over master.

kocolosk committed:
Allow configurable buffer size for all_docs, views

Now that couch_mrview can buffer chunked responses we should pass the
config value for the buffer size to it.

kocolosk committed:
Define an accessor for the chunk buffer config val

kocolosk committed:
Remove support for CouchDB =< 1.0.0

Thise UA-sniffing code has been in place to ensure interop with clients
running CouchDB versions older than 1.0.1. We should be safe to remove
it now.

kocolosk committed:
Match old behavior when buffering is disabled

If the threshold is set to zero we make sure to send a separate chunk to
close the JSON, otherwise we just merge it into the previous chunk.

kocolosk commented:
As I happened to be in this part of the codebase I took the opportunity to remove the `is_old_couch` UA-sniffing hack. It means that a replication mediated by a server running a CouchDB older than version 1.0.1 will not be able to use 2.0 as a source. Mediating the replication on the 2.0 box should still work, though honestly I don't know that anyone is actively testing this part of the compatibility matrix.

kxepal commented on src/chttpd.erl:
```diff
@@ -994,3 +1007,15 @@ stack_hash(Stack) ->
+%% response until it reaches this threshold and then send all the rows in one
+%% chunk to improve network efficiency. The default value is chosen so that
+%% the assembled chunk fits into the default Ethernet frame size (some reserved
+%% padding is necessary to accomodate the reporting of the chunk length). Set
```
Typo in "accommodate"?

kocolosk replied:
derp, thanks

kxepal commented:
@kocolosk I wonder who still uses pre 1.0 CouchDB, but if you're interested in results, I have such on vm backups to test with 2.0.

kocolosk commented:
@kxepal appreciate the offer, but personally I'd be more interested in validating replication interop as an extended part of our CI setup.

kxepal commented:
@kocolosk here cannot help you, that's another big story.

kocolosk commented:
Did a little bit of commit squashing but no changes to the final diff. Going to merge now, thanks all for the review!
